torch/csrc/autograd/variable.h

```diff
@@ -1,150 +1,250 @@
+  /// if it still exists. If the gradient accumulator function has been
+  /// destroyed, returns a `nullptr`.
+  std::shared_ptr<Function> try_get_grad_accumulator() const;
 
+  /// Get the gradient accumulator of the `Variable` if it has one, or else
+  /// create one on the fly and return it.
+  std::shared_ptr<Function> grad_accumulator() const;
+
+  /// Set the gradient edge -- i.e. `grad_fn` and `input_nr` -- of the
+  /// `Variable`.
+  /// NOTE: This will always set the `grad_fn`, even if this is a leaf
+  /// variable, and never the `grad_accumulator`. For the latter, use
+  /// `set_grad_accumulator`. This allows late construction of an interior
+  /// `Variable`.
+  void set_gradient_edge(Edge&& edge) noexcept;
+
+  /// Return the "canonical" gradient edge of this `Variable`, i.e. either the
+  /// gradient function if this is an interior `Variable`, or the gradient
+  /// accumulator otherwise. If the `Variable` is interior, the returned `Edge`
+  /// will store the input index of the `Function` to which this variable is
+  /// connected in its `input_nr` field. For leaves, the `input_nr` is always
+  /// zero. Note that `set_gradient_edge` and `gradient_edge` are not
+  /// symmetric. You must use `set_gradient_edge` to set the `grad_fn` and
+  /// `set_grad_accumulator` to set the accumulator.
   Edge gradient_edge() const {
     // If grad_fn is null (as is the case for a leaf node), we instead
-    // interpret the gradient function to be a grad accumulator,
-    // which will accumulate its inputs into the grad property of the
-    // variable. These nodes get suppressed in some situations,
-    // see "suppress grad accumulation" below. Note that only variables which
-    // have `requires_grad = True` can have grad accumulators.
+    // interpret the gradient function to be a gradient accumulator, which will
+    // accumulate its inputs into the grad property of the variable. These
+    // nodes get suppressed in some situations, see "suppress gradient
+    // accumulation" below. Note that only variables which have `requires_grad =
+    // True` can have gradient accumulators.
     if (const auto& gradient = grad_fn()) {
       return Edge(gradient, output_nr());
     } else {
       return Edge(grad_accumulator(), 0);
     }
   }
 
-  inline VariableImpl* get() const;
+  /// Return the input index of the gradient `Function` to which this `Variable`
+  /// is connected.
+  int output_nr() const noexcept;
 
-  inline const Tensor & data() const;
-  inline       Tensor & data();
+  void set_requires_grad(bool requires_grad) noexcept;
+  bool requires_grad() const noexcept;
 
-  inline Tensor opt_data() const;
+  PyObject* pyobj() const noexcept;
+  void set_pyobj(PyObject* pyobj) noexcept;
 
-  inline const Variable & grad() const;
-  inline       Variable & grad();
+  /// Set the type of the underlying `Tensor`.
+  void set_type(at::Type*) noexcept;
```

Comment:
I noticed this in code review because previously there was no setter for type, and now there is. That should have indicated that something is up.

In fact, the back story is that `set_type` is exposed *solely* to support type-changing of Variable (documented at:

```
// XXX: This is a hack to access private TensorImpl::type_
// http://bloglitb.blogspot.com/2011/12/access-to-private-members-safer.html
// This is currently needed because module.float() changes the type of the
// data field of each variable. We should fix this and not allow changing the
// type of var.data.
```

So, there should (1) be a very clear indication that `set_type` is really exposed for OO reasons, but not as a legitimate thing you should actually do with a Tensor (unless we revisit the design); at least prefixing it with an underscore but also a comment and (2) a clear reference to the portion of code which actually makes use of it, and the reason why (namely, to support parameter identity-preserving, inplace type casting of modules). You can use the `Note [Foo]` convention to do this.